core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -253,6 +210,68 @@ splitNewtypeCoercion coerce = case GHC.coercionKind coerce of
     (GHC.Pair lhs rhs@(GHC.splitTyConApp_maybe -> Just (GHC.unwrapNewTyCon_maybe -> Just (_, inner, _), _))) | GHC.eqType lhs inner -> Just $ Wrap lhs rhs
     _ -> Nothing
 
+-- Recursive types
+
+-- | Try to convert a type, using the given action to convert it.
+handleMaybeRecType :: (Converting m) => GHC.Name -> m PLCType -> m PLCType
+handleMaybeRecType n act = do
+    tds <- gets csTypeDefs
+    case Map.lookup n tds of
+        Just (InProgress tyname) -> pure $ PLC.TyVar () tyname
+        -- Either already seen and done or not seen.
+        -- Ideally we would store the finished type, but we actually need to
+        -- store a version in Quote so we can recreate it safely in multiple
+        _ -> do
+            tyname <- convTyNameFresh n
+            modify $ \cs -> cs { csTypeDefs = Map.insert n (InProgress tyname) (csTypeDefs cs)}
+            converted <- act
+            modify $ \cs -> cs { csTypeDefs = Map.insert n Done (csTypeDefs cs)}
+            pure converted
+
+{- Note [Detecting recursive types in GHC]
+This seems to be surprisingly difficult, and as far as I can tell there is nothing built in to
+do this.
+
+We can handle the self-recursive case easily enough, since we just look for the very name of
+the type constructor we're converting.
+
+For mutually recursive types we'll need to do a SCC analysis on the type dependency graph. We
+can do this as we go, but we'll want to cache it. At the moment getting the information to
+do this out of GHC seems hard.
+-}
+
+isSelfRecursiveTyCon :: (Converting m) => GHC.TyCon -> m Bool
+isSelfRecursiveTyCon tc = do
+    -- See Note [Detecting recursive types in GHC]
+    dcs <- getDataCons tc
+    let usedTcs = GHC.unionManyUniqSets $ (\dc -> GHC.unionManyUniqSets $ GHC.tyConsOfType <$> GHC.dataConRepArgTys dc) <$> dcs
+    pure $ GHC.elementOfUniqSet tc usedTcs
+
+isSelfRecursiveType :: (Converting m) => GHC.Type -> m Bool
+isSelfRecursiveType t = case t of
+    -- the only way it can be *self* recursive is if it's type constructor
+    (GHC.splitTyConApp_maybe -> Just (tc, _)) -> isSelfRecursiveTyCon tc
+    _                                         -> pure False
+
+-- | Split out a recursive type (constructed by us) into its pattern functor and the name of the variable.
+-- This is a bit of a hack, since it relies on us knowing that the structure will be some number of applications
+-- surrounding a fix.
```

Comment:
Maybe we should add elimination contexts to the AST? The fact that they do not affect syntax doesn't mean they shouldn't affect the AST as well.

Reply:
Because the head of an elimination context is always a `TyFix` and it's better to express this invariant via the AST itself. I'm pretty sure that the formalization will do this (unless elimination contexts complicate the formalization a lot and we choose something else). On the Haskell side we can have a less structured AST, but it maybe be better to have the one that is more correct-by-construction, because it's just a bit of syntactic overhead here. But I'm not sure.

Reply:
Seems worth a try.